### PR TITLE
ORC-742: Added Lazy IO for non filter columns

### PR DESCRIPTION
**Pull request by pavibhai**
### What changes were proposed in this pull request?
    * Identify columns in the presence of a filter into LEAD and FOLLOW columns
            * LEAD columns are read first
            * FOLLOW columns are read only if the filter selects an output
    * RecordReaderImpl.nextBatch performs read until a batch has value or the file is exhaused instead of returning empty batches as was the case previously
    * IO of FOLLOW columns happens at the level of RowGroup
    * In the presence of filters batches respected row group boundaries

### Why are the changes needed?
The code changes allow for a lazy evaluation of FOLLOW columns, which in the case of reads with minimal hits gives substantial savings both of IO and CPU.

### How was this patch tested?
This patch includes Unit tests that verify the IO savings accomplished as a result of this change.

### CODE REVIEW

**Comment by pavibhai** [comment]
@pgaref @omalley 
I have not broken this PR down any further so as to keep a merge functionally complete as compared to just structural changes of adding a read level for example.

If it helps with the review I am happy to break the PR into multiple commits. Please let me know.

**Comment by pavibhai** [comment]
@pgaref Following up on this to bring to your attention. If you get sometime please go through the same and provide your feedback.

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java [hunk 1]
```diff
@@ -230,15 +247,22 @@ protected RecordReaderImpl(ReaderImpl fileReader,
     SortedSet<Integer> filterColIds = new TreeSet<>();
     if (options.getPreFilterColumnNames() != null) {
       for (String colName : options.getPreFilterColumnNames()) {
-        int expandColId = findColumns(evolution, colName);
+        TypeDescription expandCol = findColumnType(evolution, colName);
         // If the column is not present in the file then this can be ignored from read.
-        if (expandColId != -1) {
-          filterColIds.add(expandColId);
+        while(expandCol != null && expandCol.getId() != -1) {
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:252]: missing space

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java [hunk 2]
```diff
@@ -958,19 +981,21 @@ public SargApplier(SearchArgument sarg,
       filterColumns = mapSargColumnsToOrcInternalColIdx(sargLeaves,
                                                         evolution);
       this.rowIndexStride = rowIndexStride;
+      this.evolution = evolution;
+      exceptionCount = new long[sargLeaves.size()];
+      this.useUTCTimestamp = useUTCTimestamp;
+    }
+
+    public void setRowIndexCols(boolean[] rowIndexCols) {
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:989]: Is this change related? Shall we move it a separate ticket if thats the case?

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java [hunk 3]
```diff
@@ -1302,25 +1427,33 @@ private int findStripe(long rowNumber) {
...
         Arrays.fill(included, true);
       }
       copy.parseStripe(stripes.get(stripeIndex), included);
-      return copy.readRowIndex(sargColumns, null);
+      return copy.readRowIndex(readCols, null);
     }
   }
 
-  private void seekToRowEntry(BatchReader reader, int rowEntry)
+  private void seekToRowEntry(BatchReader reader, int rowEntry, EnumSet<TypeReader.ReadLevel> readLevel)
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:1456]: We already have readLevel as part of our RecordReaderIml instance, any reason we also pass it as param here?

File: java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java [hunk 4]
```diff
@@ -96,16 +99,22 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
...
     }
+    int rowIdx = batch.selectedInUse ? batch.selected[rowInBatch] : rowInBatch;
     if (schema.getCategory() == TypeDescription.Category.STRUCT) {
       OrcStruct result = (OrcStruct) row;
       List<TypeDescription> children = schema.getChildren();
       int numberOfChildren = children.size();
       for(int i=0; i < numberOfChildren; ++i) {
-        result.setFieldValue(i, OrcMapredRecordReader.nextValue(batch.cols[i], rowInBatch,
-            children.get(i), result.getFieldValue(i)));
+        TypeDescription child = children.get(i);
```
Comment (pgaref) [inline java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java:108]: unrelated change?

File: java/mapreduce/src/java/org/apache/orc/mapred/OrcMapredRecordReader.java [hunk 5]
```diff
@@ -98,16 +101,22 @@ public boolean next(NullWritable key, V value) throws IOException {
...
     }
+    int rowIdx = batch.selectedInUse ? batch.selected[rowInBatch] : rowInBatch;
     if (schema.getCategory() == TypeDescription.Category.STRUCT) {
       OrcStruct result = (OrcStruct) value;
       List<TypeDescription> children = schema.getChildren();
       int numberOfChildren = children.size();
       for(int i=0; i < numberOfChildren; ++i) {
-        result.setFieldValue(i, nextValue(batch.cols[i], rowInBatch,
-            children.get(i), result.getFieldValue(i)));
+        TypeDescription child = children.get(i);
```
Comment (pgaref) [inline java/mapreduce/src/java/org/apache/orc/mapred/OrcMapredRecordReader.java:110]: unrelated change?

File: java/core/src/test/org/apache/orc/impl/TestSchemaEvolution.java [hunk 6]
```diff
@@ -1732,10 +1732,10 @@ public void testTypeConversion() throws IOException {
         dataReader, OrcFile.WriterVersion.ORC_14, true, Integer.MAX_VALUE);
     boolean[] columns = new boolean[]{true, true, true};
     planner.parseStripe(dataReader.getStripe(0), columns)
-        .readData(null, null, false);
-    reader.startStripe(planner);
+        .readData(null, null, false, TypeReader.ReadLevel.ALL);
+    reader.startStripe(planner, TypeReader.ReadLevel.ALL);
     VectorizedRowBatch batch = readType.createRowBatch();
-    reader.nextBatch(batch, 10);
+    reader.nextBatch(batch, 10, TypeReader.ReadLevel.ALL);
```
Comment (pgaref) [inline java/core/src/test/org/apache/orc/impl/TestSchemaEvolution.java:1738]: Use overloaded method for all and create a separate test for partial?

File: java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java [hunk 7]
```diff
@@ -144,10 +165,23 @@ public StripePlanner parseStripe(StripeInformation stripe,
    * @return the buffers that were read
    */
   public BufferChunkList readData(OrcIndex index,
-                                boolean[] rowGroupInclude,
-                                boolean forceDirect) throws IOException {
+                                  boolean[] rowGroupInclude,
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java:168]: Update doc here? how readLevel is changing the behaviour of the method?

File: java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java [hunk 8]
```diff
@@ -24,22 +25,51 @@
...
+   * @param readLevel the requested read level
+   * @return true if allowed by read level or if it is a LEAD_PARENT otherwise false
+   */
+  static boolean allowChild(TypeReader reader, EnumSet<ReadLevel> readLevel) {
+    return readLevel.contains(reader.getReadLevel())
+           || reader.getReadLevel() == ReadLevel.LEAD_PARENT;
+  }
+
+  enum ReadLevel {
+    LEAD_CHILD,    // Read only the elementary filter columns
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java:65]: Do we really need to distinguish between LEAD_CHILD and LEAD_PARENT? 
Why not just LEAD and FOLLOW?

**Review by pgaref — CHANGES_REQUESTED** [review]
Hey @pavibhai thanks for pinging me! 
This is much cleaner than before but still hard to parse as it contains many critical changes.

Left some preliminary comments but my general advice would be to have one final split of this where we just introduce the enums and prepare the Reader for the actual planning/reading changes -- for instance I am sure we can cut down on the places where we pass ReadLevel around, and introduce some overloaded method where necessary.

In the final PR we will then only need to introduce the logic of partialPlan and follow Data Read.

Please let me know what you think

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java (hunk 1 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:252]: will add

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java (hunk 3 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:1456]: This is still needed as we do two types of seeks within RecordReader one for just the filter columns and the other that performs only the non-filter columns when the RecordReader readLevel says filter columns only.

File: java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java (hunk 4 above)
Comment (pavibhai) [reply java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java:108]: Sorry, yes this should have been a separate bug fix patch that MapRed reader does not respect inclusions when setting the values.

Will pull this out.

File: java/mapreduce/src/java/org/apache/orc/mapred/OrcMapredRecordReader.java (hunk 5 above)
Comment (pavibhai) [reply java/mapreduce/src/java/org/apache/orc/mapred/OrcMapredRecordReader.java:110]: Sorry, yes this should have been a separate bug fix patch that MapRed reader does not respect inclusions when setting the values.

Will pull this out.

File: java/core/src/test/org/apache/orc/impl/TestSchemaEvolution.java (hunk 6 above)
Comment (pavibhai) [reply java/core/src/test/org/apache/orc/impl/TestSchemaEvolution.java:1738]: To be clear you are asking can we add overloaded methods on the reader that if the level is not passed it will be treated as ALL. Please confirm.

I think that method will only be used from test, in all other cases the reader level is always supplied by the RecordReader.

File: java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java (hunk 7 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java:168]: Will add

File: java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java (hunk 8 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java:65]: The LEAD_PARENT has significance that it should still propagate activities on the child even if it does not satisfy the requested read level. That is why I chose to differentiate between PARENT and CHILD.

**Comment by pavibhai** [comment]
> Left some preliminary comments but my general advice would be to have one final split of this where we just introduce the enums and prepare the Readers for the actual planning/reading changes -- for instance I am sure we can cut down on the places where we pass ReadLevel around, and introduce some overloaded method where necessary.
> 
> Then, in the final PR we will then only need to introduce the logic of partialPlan and follow Data Read which will be much cleaner.
> 
> Please let me know what you think :)

@pgaref Thanks for your feedback. Instead of separating PRs as the PR with just the level changes as it does not add any functionality, maybe alternatively shall I break the PR into multiple commits so as to differentiate these changes:
1. Introduce ReadLevel and add the parameters to the methods
2. Introduce the LazyIO based on the read level

Let me know what you think.

File: java/core/src/test/org/apache/orc/impl/TestSchemaEvolution.java (hunk 6 above)
Comment (pgaref) [reply java/core/src/test/org/apache/orc/impl/TestSchemaEvolution.java:1738]: If its only used by the test then lets leave it as

File: java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java (hunk 8 above)
Comment (pgaref) [reply java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java:65]: Not sure I follow, in general every column is a child of the root struct in ORc right?
 Can you share an example please where we need the distinction?

**Comment by pgaref** [comment]
> > Left some preliminary comments but my general advice would be to have one final split of this where we just introduce the enums and prepare the Readers for the actual planning/reading changes -- for instance I am sure we can cut down on the places where we pass ReadLevel around, and introduce some overloaded method where necessary.
> > Then, in the final PR we will then only need to introduce the logic of partialPlan and follow Data Read which will be much cleaner.
> > Please let me know what you think :)
> 
> @pgaref Thanks for your feedback. Instead of separating PRs as the PR with just the level changes does not add any functionality, maybe alternatively shall I break the PR into multiple commits so as to differentiate these changes:
> 
> 1. Introduce ReadLevel and add the parameters to the methods
> 2. Introduce the LazyIO based on the read level
> 
> Let me know what you think.

Hey @pavibhai that could also help, see what works better for you :)

File: java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java (hunk 8 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java:65]: Sure.

Consider the following schema and classification
* f1: Long
* f2: String
* f3: Struct
  * a: String
  * b: Long

In the above schema if the filter was based on f2 and f3.a then the classification will be as follows:
* f1: Long: FOLLOW
* f2: String: LEAD_CHILD
* f3: Struct: LEAD_PARENT
  * a: String: LEAD_CHILD
  * b: Long: FOLLOW

The difference of LEAD_PARENT comes into play from the standpoint of seeks. Where a seek for follow involves the seek of the LEAD_PARENTs as well to determine the non null rows in the reader.

This is documented in [prepareFollowReaders](https://github.com/apache/orc/blob/cc48b40e245f717ffc79c170f94f750da4fccf32/java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java#L1305)

This will come into play when let us say the lead read has already happened to say Row 8123 while the previous follow read has happened to say 1000. Now we want to skip to rows 8123 on the follow readers, this would require the determination of number of non null values on the parent, in our example for f3.b.

This is achieved in the following steps:
1. Seek LEAD_PARENT to follow position i.e. 1000
2. Seek LEAD_PARENT + FOLLOW to desired position 8123, this will ensure that non nulls is determined correctly and the follows are seeked.

LEAD_CHILD are unaffected and remain at 1000 through the above process, at the end all of them are on 1000. Hope this is clear. If this is better handled via a call, I am open for it.

**Comment by pgaref** [comment]
Hey @pavibhai  thanks updating this! Any remaining comments to address? Is this ready for a final review?

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java (hunk 2 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:1025]: It is related. I see that the rowIndexCols are not used within the SArgApplier. Additionally now the RowIndexCols is decided both by the SArgs and the Filters so I wanted to pull this out of the SArgApplier into RecordReaderImpl where it is used.

**Comment by pavibhai** [comment]
> Hey @pavibhai thanks updating this! Any remaining comments to address? Is this ready for a final review?

@pgaref thanks for following up. I think most of the changes are complete. The key item that I have not yet done is splitting the PR. 

I have had some discussions with @omalley and based on that the key change you will see in the PR right now is the separation of ReadLevel into ReadCategory and ReadPhase.

Hope that makes sense.

File: java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java (hunk 8 above)
Comment (pgaref) [reply java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java]: Thanks for the detailed explanation!

File: java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java [hunk 9]
```diff
@@ -24,22 +25,70 @@
...
+  void skipRows(long rows, ReadPhase readPhase) throws IOException;
 
   void nextVector(ColumnVector previous,
                   boolean[] isNull,
                   int batchSize,
-                  FilterContext filterContext) throws IOException;
+                  FilterContext filterContext,
+                  ReadPhase readPhase) throws IOException;
 
   int getColumnId();
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java:47]: not sure we need getColumnId() anymore?

File: java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java [hunk 10]
```diff
@@ -189,12 +192,34 @@ public boolean useProlepticGregorian() {
     public boolean fileUsedProlepticGregorian() {
       return fileUsedProlepticGregorian;
     }
+
+    @Override
+    public TypeReader.ReaderCategory getReaderCategory(int columnId) {
+      TypeReader.ReaderCategory result;
+      if (getColumnFilterCallback() == null || getColumnFilterIds().contains(columnId)) {
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:199]: Not sure why consider getColumnFilterCallback null case here? 
When there is not filter set we consider all ReaderCategories to be FILTER_PARENT or FILTER_CHILD?

File: java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java [hunk 11]
```diff
@@ -2856,11 +2962,16 @@ public static TypeReader createTreeReader(TypeDescription readerType,
     }
   }
 
+  public static boolean isDecimalAsLong(OrcFile.Version version, int precision) {
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:2965]: This could be part of another PR tbh

File: java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java [hunk 12]
```diff
@@ -445,55 +509,83 @@ private static boolean hasSomeRowGroups(boolean[] includedRowGroups) {
...
     if (hasSomeRowGroups(includedRowGroups)) {
       InStream.StreamOptions compression = dataReader.getCompressionOptions();
       boolean isCompressed = compression.getCodec() != null;
       int bufferSize = compression.getBufferSize();
       OrcProto.RowIndex[] rowIndex = index.getRowGroupIndex();
       for (StreamInformation stream : dataStreams) {
-        if (RecordReaderUtils.isDictionary(stream.kind, encodings[stream.column])) {
-          addChunk(result, stream, stream.offset, stream.length);
+        if (readPhase == TypeReader.ReadPhase.ALL
+            ||(readPhase == TypeReader.ReadPhase.LEADERS
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java:523]: nit. space after || 
nit. reduce new lines

File: java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java (hunk 9 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java:47]: Sure. This was not added as part of this PR. I can raise a separate PR to remove this or include the removal here.

File: java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java (hunk 10 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:199]: Good catch. I missed cleaning up this condition, I already inverted in the previous commit that in the absence of filters we treat every column as NON_FILTER. Am removing `getColumnFilterCallback() == null` condition.

File: java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java (hunk 11 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:2965]: I need this for the next PR conversion of SArg to Filter. I can remove this here and include it as part of that PR

Comment (pgaref) [reply java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:2965]: Sounds good! The SArgs PR will be smaller anyway so better to add there

File: java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java (hunk 9 above)
Comment (pgaref) [reply java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java:47]: Lets do that as part of the follow up PR -- this part of the code is not released yet (only part of main branch) so we have more flexibility

File: java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java (hunk 12 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java:523]: > nit. space after ||
done

> nit. lets keep conditions on a single line
brought the subcondition into a single line not the entire condition as it extends the line boundary significantly

> nit. reduce new lines (e.g., processStream call)
Fixed the processStream call however I am unclear on how to handle parameters. I have generally been following chompdown if long, I see both variants in the code chomp or just wrap.

File: java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java (hunk 11 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:2965]: reverted this change

File: java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java (hunk 12 above)
Comment (pgaref) [reply java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java]: Looks good now, thanks!

**Review by pgaref — APPROVED** [review]
Thanks for addressing all the comments @pavibhai ! 
+1 on the latest changes pending tests

**Comment by pavibhai** [comment]
> This is now merged to main! Thanks @pavibhai for all the hard work!

Thanks @pgaref

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java [hunk 13]
```diff
@@ -23,6 +23,7 @@
 import java.time.ZoneOffset;
 import java.time.chrono.ChronoLocalDate;
 import java.time.format.DateTimeFormatter;
+import org.apache.commons.lang.ArrayUtils;
```
Comment (dongjoon-hyun) [inline java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:26]: Although ORC-734 removed `commons-lang` usage from Apache ORC in favor of `lang3`, this seems to bring it back. I'll make a PR to remove it.

Comment (dongjoon-hyun) [reply java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:26]: Here is the PR.
- https://github.com/apache/orc/pull/811
